PER-9717 [back-end] Endpoint to retrieve environment configuration

iulianvsp opened the pull request:
PER-9717 [back-end] Endpoint to retrieve environment configuration

liam-lloyd commented on packages/api/src/account/queries/get_feature_flags.sql:
```diff
@@ -0,0 +1,6 @@
+SELECT
```
This query should live under /feature_flag/queries rather than here, since it's dealing with feature flags at the global level rather than at the account level

liam-lloyd commented on packages/api/src/feature/models.ts:
```diff
@@ -0,0 +1,11 @@
+  globallyEnabled: boolean;
+}
+
+export interface FeatureFlagRow {
+  id: number;
+  name: string;
+  globallyEnabled: boolean;
+}
```
This looks identical to the interface above; what is it for?

liam-lloyd commented on packages/api/src/feature/service.ts:
```diff
@@ -0,0 +1,32 @@
+      );
+    });
+
+  const featureFlags = result.rows.map<FeatureFlag>(
+    (row: FeatureFlagRow): FeatureFlag => ({
+      ...row,
+      id: row.id,
+      name: row.name,
+      globallyEnabled: row.globallyEnabled,
+    })
+  );
```
We use code like this in other endpoints to transform some fields where the desired output is different from what's in the database, but I don't think there's a need for that here.

liam-lloyd commented on packages/api/src/feature/controller.ts:
```diff
@@ -0,0 +1,27 @@
+      const featureFlags = await featureService.getFeatureFlags();
+      res.json(featureFlags);
+    } catch (err) {
+      if (isValidationError(err)) {
+        res.status(400).json({ error: err });
+        return;
```
Since this endpoint doesn't make use of any inputs, there is nothing to validate, so we needn't check for validation errors

liam-lloyd reviewed (commented):
You're generally on the right path here as far as how we write endpoints and what this one should do. I've added a couple comments in places where I think a better course exists.

You asked also whether there are any additional fields this endpoint should return; I've now added an [OpenAPI specification](https://github.com/PermanentOrg/stela/blob/main/packages/api/docs/present/paths/feature_flags.yaml) describing the desired endpoint in detail (though that specification need not be taken as dogma; if you see a better way by all means present your case for it). If you prefer not to read the YAML directly, you can create a much nicer HTML version of the specification by running `redocly build-docs packages/api/docs/present/api.yaml` (we have just recently adopted this API specification format and aren't yet hosting a docs site).

Some other documentation that might give you useful further context:
[The longer term plan for feature flags](https://permanent.atlassian.net/wiki/spaces/EN/pages/2221473793/Feature+Flags)
[Our best practices for API design](https://permanent.atlassian.net/wiki/spaces/EN/pages/2203451393/API+Design+Principles)

iulianvsp commented:
> You're generally on the right path here as far as how we write endpoints and what this one should do. I've added a couple comments in places where I think a better course exists.
> 
> You asked also whether there are any additional fields this endpoint should return; I've now added an [OpenAPI specification](https://github.com/PermanentOrg/stela/blob/main/packages/api/docs/present/paths/feature_flags.yaml) describing the desired endpoint in detail (though that specification need not be taken as dogma; if you see a better way by all means present your case for it). If you prefer not to read the YAML directly, you can create a much nicer HTML version of the specification by running `redocly build-docs packages/api/docs/present/api.yaml` (we have just recently adopted this API specification format and aren't yet hosting a docs site).
> 
> Some other documentation that might give you useful further context: [The longer term plan for feature flags](https://permanent.atlassian.net/wiki/spaces/EN/pages/2221473793/Feature+Flags) [Our best practices for API design](https://permanent.atlassian.net/wiki/spaces/EN/pages/2203451393/API+Design+Principles)

I updated the PR with the suggested changes, please have a look. I also saw in the best practices API design page that for lists of object we want to have pagination. Does this apply to the feature_flags as well?

Also should I create the other CRUD methods to this entity? Thanks

liam-lloyd commented:
> I also saw in the best practices API design page that for lists of object we want to have pagination. Does this apply to the feature_flags as well?

I think feature flags are most naturally an exception to this rule. From the user-facing client's perspective, it would be a hassle to have to make several calls through multiple pages of feature flags in order to determine which features should be active. From the admin panel's perspective, pagination might be more reasonable, but I think realistically we should never have so many feature flags at once that they would be unmanageable without pagination (unless we neglect to delete old ones, but then the lack of pagination may become a useful inducement to not neglect that).

> Also should I create the other CRUD methods to this entity?

I think that would be a great thing for you to do once this endpoint is done; the create and update methods are higher priority than delete. The tickets for those endpoints are here:
https://permanent.atlassian.net/browse/PER-9718
https://permanent.atlassian.net/browse/PER-9866

iulianvsp commented on packages/api/src/feature_flag/controller.ts:
```diff
@@ -0,0 +1,21 @@
+
+featureController.get(
+  "/features",
+  verifyUserAuthentication,
```
I created extractUserIsAdminFromAuthToken, but please have a look at my comments there. I'm not sure if i did that correctly

liam-lloyd commented on packages/api/src/middleware/authentication.ts:
```diff
@@ -228,9 +228,33 @@ const extractUserEmailFromAuthToken = async (
+      req.body.admin = true;
+    }
+    next();
+  } catch (innerErr) {
+  }
```
This empty catch doesn't seem like the right path here; in almost any case when an exception occurs we at least want to know about it. If this is what fixed the timeout error, the error is presumably still happening, it just gets swallowed by this catch now.

Since what you're doing here is just about the same as what `extractUserEmailFromAuthToken` does, I suspect what's going on is that your FusionAuth env vars aren't set to the correct values; indeed it looks like they couldn't be, because you don't have access to FusionAuth yet. I'll set up your account there now.

liam-lloyd replied:
You should have an email inviting you to FusionAuth; once you get into their admin console you should be able to find the values you need for the environment variables.

liam-lloyd commented on packages/api/src/idpuser/models.ts:
```diff
@@ -36,3 +36,7 @@ export enum TwoFactorMethod {
   Sms = "sms",
 }
+
+export interface FeatureFlagRequest {
+  admin: boolean;
+}
```
Shouldn't this be in the `feature_flag` models?

liam-lloyd commented on packages/api/src/feature_flag/controller.test.ts:
```diff
@@ -0,0 +1,79 @@
+
+jest.mock("../database");
+jest.mock("../middleware");
+jest.mock("../mixpanel");
```
I don't think these tests should require mocking mixpanel

liam-lloyd commented on packages/api/src/feature_flag/controller.test.ts:
```diff
@@ -0,0 +1,79 @@
+jest.mock("../middleware");
+jest.mock("../mixpanel");
+
+describe("GET /feature-flags", () => {
```
These tests look great, but it's a good idea to also test the error cases to make sure that we send the right status code and log the error correctly if, say, the database call fails.

liam-lloyd replied:
I'd add two more error cases: One where the database call fails when the caller is an admin, and one where the request is invalid.

liam-lloyd commented on packages/api/src/feature_flag/controller.ts:
```diff
@@ -0,0 +1,22 @@
+  extractUserIsAdminFromAuthToken,
+  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+    try {
+      validateIsAdminFromAuthentication(req.body);
```
Now that we have a validator in this controller, we should add back the check for validation errors (and write a test that we return the correct status code in case of a validation error).

liam-lloyd commented on packages/api/src/middleware/authentication.ts:
```diff
@@ -228,9 +228,25 @@ const extractUserEmailFromAuthToken = async (
   next();
 };
 
+const extractUserIsAdminFromAuthToken = async (
```
You asked about unit tests and I think my answer was incomplete. To answer more thoroughly, we favor functional tests for our controllers (and also, so far, for lambda handlers and cron scripts), because we want to ensure we test the SQL against a real database, and because separating the service from the controller in tests doesn't seem to add much value. We do generally unit test code that is shared between many controllers, such as middleware or shared validators.

liam-lloyd replied:
It looks like this function still needs unit tests

liam-lloyd commented on packages/api/src/feature_flag/controller.test.ts:
```diff
@@ -0,0 +1,112 @@
+      }
+    );
+
+    await agent.get("/api/v2/feature-flags").expect(500);
```
For an invalid request we should expect a 400 error

iulianvsp replied:
This scenario would not be possible because extractUserIsAdminFromAuthToken would always return a boolean request.body.admin value. Maybe it would be best to just remove this test?

iulianvsp replied:
Done, thanks :)

liam-lloyd reviewed (commented):
Just one more comment, then I think this will be ready to merge!

liam-lloyd commented on packages/api/src/middleware/authentication.ts:
```diff
@@ -228,9 +228,25 @@ const extractUserEmailFromAuthToken = async (
+  next: NextFunction
+): Promise<void> => {
+  const authenticationToken = getAuthTokenFromRequest(req);
+  const email = await getOptionalValueFromAuthToken(
```
I found one more thing while testing locally - this function will throw an exception (because of the way the Fusionauth client works) if it receives an `authenticationToken` that's an empty string. Since we don't want that case to fail, we should check for an empty `authenticationToken` and then only call this if it's not empty.
